src/main/java/com/google/sps/servlets/addSchoolInfo.java

```diff
@@ -0,0 +1,171 @@
+      newCourse.setProperty("course-name", courseName);
+      newCourse.setProperty("Units", units);
+      newCourseKey = newCourse.getKey();
+      datastore.put(newCourse);
+
+      Entity newProfessor;
+      newProfessor = new Entity("Professor", existingSchoolKey);
+      newProfessor.setProperty("professor-name", profName);
+      newProfessorKey = newProfessor.getKey();
+      datastore.put(newProfessor);
+
+      Entity newTerm = new Entity("Term", newCourseKey);
+      newTerm.setProperty("term", term);
+      newTerm.setProperty("Professor", newProfessorKey);
+      datastore.put(newTerm);
+
+    } else {
+      Filter courseFilter = new FilterPredicate("course-name", FilterOperator.EQUAL, courseName);
+      Query courseQ = new Query("Course").setAncestor(existingSchoolKey).setFilter(courseFilter);
+      List<Entity> courseQuery =
+          datastore.prepare(courseQ).asList(FetchOptions.Builder.withDefaults());
+
+      // existing school --> new course
+      if (courseQuery.size() == 0) {
+        Entity newCourse;
+        if (existingSchoolKey != null) {
+          newCourse = new Entity("Course", existingSchoolKey);
+        } else {
+          newCourse = new Entity("Course", newSchoolKey);
+        }
+
+        newCourse.setProperty("course-name", courseName);
+        newCourse.setProperty("Units", units);
+        datastore.put(newCourse);
+
+        Entity newTerm = new Entity("Term", newCourseKey);
+        newTerm.setProperty("term", term);
+
+        Filter professorFilter =
+            new FilterPredicate("professor-name", FilterOperator.EQUAL, profName);
+        Query professorQ =
+            new Query("Professor").setAncestor(existingSchoolKey).setFilter(professorFilter);
+        List<Entity> professorQuery =
+            datastore.prepare(professorQ).asList(FetchOptions.Builder.withDefaults());
+
+        // existing school --> new course --> new professor
+        if (professorQuery.size() == 0) {
+          System.out.println("OLD SCHOOL NEW COURSE NEW PROFESSOR");
+          Entity newProfessor;
+          newProfessor = new Entity("Professor", existingSchoolKey);
+          newProfessor.setProperty("professor-name", profName);
+          newProfessorKey = newProfessor.getKey();
+          datastore.put(newProfessor);
+
+          newTerm.setProperty("Professor", newProfessorKey);
+          datastore.put(newTerm);
+
+        } else {
+          // existing school --> new course --> existing professor
+          Entity existingProfessor = professorQuery.get(0);
+          existingProfessorKey = existingProfessor.getKey();
+
+          newTerm.setProperty("Professor", newProfessorKey);
+          datastore.put(newTerm);
+        }
+      } else {
+        // existing school --> existing course --> existing professor
+        Entity existingCourse = courseQuery.get(0);
+        existingCourseKey = existingCourse.getKey();
+
+        Entity newTerm = new Entity("Term", existingCourseKey);
```

Comment:
This would be an existing term right? (consider changing the variable name maybe)

Reply:
It's actually a new term in all cases because we are assuming you can't have duplicate terms. I didn't make that check, but for safety I will change it such that if a term already exists we won't make a new one.

Reply:
Perfect! :)